### PR TITLE
update TRDH to use diagonal QN model

### PR DESCRIPTION
**Pull request by dpo**

### CODE REVIEW

File: src/TR_alg.jl [hunk 1]
```diff
@@ -186,8 +186,9 @@ function TR(
     set_radius!(ψ, ∆_effective)
     subsolver_options.Δk = ∆_effective / 10
     subsolver_options.ν = ν
+    subsolver_args = subsolver == TRDH ? (SpectralGradient(1 / ν, f.meta.nvar),) : ()
```
Comment (geoffroyleconte) [inline src/TR_alg.jl:189]: Does this imply that we can only use TR-TRDH with `SpectralGradient`? 
Also, I think the operator should be preallocated outside the while loop to facilitate the use of in-place subsolvers (once in-place methods are implemented for all the subsolvers in the same manner as `R2!`)

**Review by geoffroyleconte — CHANGES_REQUESTED** [review]

File: src/TR_alg.jl (hunk 1 above)
Comment (dpo) [reply src/TR_alg.jl:189]: Yes, for the moment, it’s the only option (the other two don’t perform well anyways). I plan to make this an option when I implement the solver object, so I’m not making an effort to preallocate at this point.

Comment (geoffroyleconte) [reply src/TR_alg.jl:189]: The others perform well on some problems, like SVM for example. I'll keep a branch on my fork then if want still want to be able to produce the results of indef-pg.

Comment (dpo) [reply src/TR_alg.jl:189]: I’m planning to work on the solvers soon, so the option should be in place in the near future.

Comment (dpo) [reply src/TR_alg.jl:189]: I added a note about that in #129.

**Pull request merged**